hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1066 @@
+               const xed_encoder_operand_t& op_1,
+               const xed_encoder_operand_t& op_2,
+               const xed_encoder_operand_t& op_3,
+               xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, op_1, op_2, op_3, effOperandSizeBits,
+                                   dest()));
+  }
+
+public:
+#define INT3_SIZE   (sz::byte)
+#define UD2_SIZE    (sz::word)
+
+  void emitInt3s(int n) {
+    static uint8_t instr = 0;
+    if (n == 0) return;
+    if (instr == 0) {
+      xedEmit(XED_ICLASS_INT3, sz::byte, &instr);
+    }
+    for (auto i = 0; i < n; ++i) {
+      byte(instr);
+    }
+  }
+
+  void emitNop(int n) {
+    if (n == 0) return;
+    static const xed_iclass_enum_t nops[] = {
+      XED_ICLASS_INVALID,
+      XED_ICLASS_NOP,
+      XED_ICLASS_NOP2,
+      XED_ICLASS_NOP3,
+      XED_ICLASS_NOP4,
+      XED_ICLASS_NOP5,
+      XED_ICLASS_NOP6,
+      XED_ICLASS_NOP7,
+      XED_ICLASS_NOP8,
+      XED_ICLASS_NOP9,
+    };
+    // While n >= 9, emit 9 byte NOPs
+    while (n >= 9) {
+      xedInstr(XED_ICLASS_NOP9, 0);
+      n -= 9;
+    }
+    // Emit remaining NOPs (if any)
+    if (n) {
+      xedInstr(nops[n], 0);
+    }
+  }
+
+  void pad() {
+    auto remaining = available();
+    static uint8_t padInstructs[UD2_SIZE + INT3_SIZE] = {0};
+    if (padInstructs[0] == 0) {
+      xedEmit(XED_ICLASS_UD2, 0, padInstructs);
+      xedEmit(XED_ICLASS_INT3, 0, padInstructs + UD2_SIZE);
+    }
+    while (remaining >= UD2_SIZE) {
+      bytes(UD2_SIZE, padInstructs);
+      remaining -= UD2_SIZE;
+    }
+    while (remaining >= INT3_SIZE) {
+      byte(padInstructs[UD2_SIZE]);
+      remaining -= INT3_SIZE;
+    }
+  }
+
+public:
+  /*
+   * The following functions use a naming convention for an older API
+   * to the assembler; conditional loads and moves haven't yet been
+   * ported.
+   */
+
+  // CMOVcc [rbase + off], rdest
+  inline void cload_reg64_disp_reg64(ConditionCode cc, Reg64 rbase,
+                                     int off, Reg64 rdest) {
+    MemoryRef m(DispReg(rbase, off));
+    xedInstrMR(ccToXedCMov(cc), m, rdest);
+  }
+  inline void cload_reg64_disp_reg32(ConditionCode cc, Reg64 rbase,
+                                     int off, Reg32 rdest) {
+    MemoryRef m(DispReg(rbase, off));
+    xedInstrMR(ccToXedCMov(cc), m, rdest);
+  }
+  inline void cmov_reg64_reg64(ConditionCode cc, Reg64 rsrc, Reg64 rdest) {
+    xedInstrRR(ccToXedCMov(cc), rsrc, rdest);
+  }
+
+  XedAssembler& fs(MemoryRef* mr)  { mr->setSegment(int(XED_REG_FS));
+                                     return *this; }
+  XedAssembler& gs(MemoryRef* mr)  { mr->setSegment(int(XED_REG_GS));
+                                     return *this; }
+
+private:
+  /*
+   * The following section contains conversion methods that take a Reg8/32/64,
+   * RegXMM, MemoryRef, RipRelative struct and convert it to a
+   * xed_encoder_operand_t.
+   */
+
+  static constexpr int sizeToBits(int sz) {
```

Comment:
nit: maybe call this "bytesToBits" to be clearer

Reply:
Addressed in https://github.com/facebook/hhvm/pull/8164/commits/ffa4555156393d19cfa7b8b0ca6f51db5253dae2